### ISSUE TITLE
"CuArray not Defined" because CUDA is not detected by `check_compiler` on Windows

### ISSUE THREAD

**Issue by mingels28** [WaterLily-jl/WaterLily-Benchmarks#3]
I am trying to benchmark a few different systems and the CPU benchmark works great, but when I try to benchmark with a GPU, I get a consistent error:

``` Julia
ERROR: LoadError: UndefVarError: `CuArray` not defined
```

I have tried several different calls to ```benchmark.sh``` as well as adding ```using CUDA``` to the top of ```benchmark.jl``` and the issue still persists. I have tried this on Windows (WSL), Ubuntu, and on an HPC cluster.

The calls to ```benchmark.sh``` I have tried are as follows:

1.) ```bash ./benchmark.sh```

2.) ```bash ./benchmark.sh -wd "../WaterLily.jl" -w "" -v "release" -t "1 4" -b "Array CuArray" -c "tgv jelly" -p "6,7 5,6" -s "100 100" -ft "Float32 Float32"```

3.) ```bash ./benchmark.sh -wd "../WaterLily.jl" -w "" -v "release" -b "CuArray" -c "tgv jelly" -p "6,7 5,6" -s "100 100" -ft "Float32 Float32"```

**Comment by b-fg** [comment]
Hi! Thanks for reporting this. Running the GPU benchmarks with `sh benchmarks.sh -b "CuArray"` I cannot reproduce this error, so could you please include the full error log?

**Comment by mingels28** [comment]
Sure thing! Here is the full error log:

```
ERROR: LoadError: UndefVarError: `CuArray` not defined
Stacktrace:
 [1] top-level scope
   @ :0
 [2] eval
   @ ./boot.jl:385 [inlined]
 [3] eval
   @ ./client.jl:491 [inlined]
 [4] #2
   @ /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/util.jl:13 [inlined]
 [5] |>
   @ ./operators.jl:917 [inlined]
 [6] parse_cla(args::Vector{String}; cases::Vector{String}, log2p::Vector{Tuple{Int64, Int64}}, max_steps::Vector{Int64}, ftype::Vector{DataType}, backend::Type)
   @ Main /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/util.jl:13
 [7] top-level scope
   @ /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:80
in expression starting at /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:80
```

which I got by calling:
```
bash ./benchmark.sh -wd "../WaterLily.jl" -v "release" -b "CuArray" -c "tgv jelly" -p "6,7 5,6" -s "100 100" -ft "Float32 Float32"
```

**Comment by b-fg** [comment]
I have different line numbers. Can you please do a `git fetch; git pull`, run the same, and report back?

**Comment by mingels28** [comment]
No problem, ran ```git fetch; git pull``` and here is the error message

```
ERROR: LoadError: UndefVarError: `CuArray` not defined
Stacktrace:
 [1] top-level scope
   @ :0
 [2] eval
   @ ./boot.jl:385 [inlined]
 [3] eval
   @ ./client.jl:491 [inlined]
 [4] #2
   @ /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/util.jl:13 [inlined]
 [5] |>
   @ ./operators.jl:917 [inlined]
 [6] parse_cla(args::Vector{String}; cases::Vector{String}, log2p::Vector{Tuple{Int64, Int64}}, max_steps::Vector{Int64}, ftype::Vector{DataType}, backend::Type, data_dir::String)
   @ Main /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/util.jl:13
 [7] top-level scope
   @ /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:81
in expression starting at /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:81
```

differing line numbers were likely from where I added ```using CUDA``` at the top

**Comment by b-fg** [comment]
So in util.jl, there are a few lines to check on wether CUDA is installed or not:
```
backend_str = Dict(Array => "CPUx"*@sprintf("%.2d", Threads.nthreads()))
check_compiler(compiler,parse_str) = try occursin(parse_str, read(`$compiler --version`, String)) catch _ false end
_cuda = check_compiler("nvcc","release")
_cuda && (using CUDA: CuArray; backend_str[CuArray] = "CUDA")
```
Can you try to run this in the REPL, using `julia --project` from the WaterLily-Benchmarks directory? Check the contents in `backend_str`.

**Comment by mingels28** [comment]
Here's what I got:

```
julia> backend_str = Dict(Array => "CPUx"*@sprintf("%.2d", Threads.nthreads()))
sin(parse_str, read(`$compiler --version`, String)) catch _ false end
_cuda = check_compiler("nvcc","release")
_cuda && (using CUDA: CuArray; backend_str[CuArray] = "CUDA")Dict{UnionAll, String} with 1 entry:
  Array => "CPUx01"

julia> check_compiler(compiler,parse_str) = try occursin(parse_str, read(`$compiler --version`, String)) catch _ false end
check_compiler (generic function with 1 method)

julia> _cuda = check_compiler("nvcc","release")
false

julia> _cuda && (using CUDA: CuArray; backend_str[CuArray] = "CUDA")
false
```

Then if I check ```backend_str```
```
Dict{UnionAll, String} with 1 entry:
  Array => "CPUx01"
  ```
Because of this, I tried adding ```nvidia-cuda-toolkit``` which gave the following:

```
julia> _cuda = check_compiler("nvcc","release")
true

julia> _cuda && (using CUDA: CuArray; backend_str[CuArray] = "CUDA")
"CUDA"

julia> backend_str
Dict{UnionAll, String} with 2 entries:
  Array   => "CPUx01"
  CuArray => "CUDA"
  ```
  
  However when trying to run the benchmark again, we get
  ```
  ERROR: LoadError: CUDA error: an illegal memory access was encountered (code 700, ERROR_ILLEGAL_ADDRESS)
  ```
  I can also add the full error log but it is much longer than before.

**Comment by b-fg** [comment]
Now that you have installed `nvidia-cuda-toolkit`, can you check the output of `CUDA.functional()`?

**Comment by mingels28** [comment]
returns ```true```. However I am pretty sure this was the output previously as well if I manually import CUDA in a Julia script.

**Comment by b-fg** [comment]
I guess we will need to look at the full error log to see why the error is showing up then. Can you post it please?

**Comment by mingels28** [comment]
```
ERROR: LoadError: CUDA error: an illegal memory access was encountered (code 700, ERROR_ILLEGAL_ADDRESS)
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:30
  [2] nonblocking_synchronize(val::CUDA.CuContext)
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/synchronization.jl:174
  [3] device_synchronize(; blocking::Bool, spin::Bool)
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/synchronization.jl:186
  [4] device_synchronize
    @ ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/synchronization.jl:180 [inlined]
  [5] checked_cuModuleLoadDataEx(_module::Base.RefValue{Ptr{CUDA.CUmod_st}}, image::Ptr{UInt8}, numOptions::Int64, options::Vector{CUDA.CUjit_option_enum}, optionValues::Vector{Ptr{Nothing}})
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/module.jl:18
  [6] CUDA.CuModule(data::Vector{UInt8}, options::Dict{CUDA.CUjit_option_enum, Any})
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/module.jl:60
  [7] CuModule
    @ ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/module.jl:49 [inlined]
  [8] link(job::GPUCompiler.CompilerJob, compiled::@NamedTuple{image::Vector{UInt8}, entry::String})
    @ CUDA ~/.julia/packages/CUDA/Tl08O/src/compiler/compilation.jl:413
  [9] actual_compilation(cache::Dict{Any, CUDA.CuFunction}, src::Core.MethodInstance, world::UInt64, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::typeof(CUDA.compile), linker::typeof(CUDA.link))
    @ GPUCompiler ~/.julia/packages/GPUCompiler/Y4hSX/src/execution.jl:257
 [10] cached_compilation(cache::Dict{Any, CUDA.CuFunction}, src::Core.MethodInstance, cfg::GPUCompiler.CompilerConfig{GPUCompiler.PTXCompilerTarget, CUDA.CUDACompilerParams}, compiler::Function, linker::Function)
    @ GPUCompiler ~/.julia/packages/GPUCompiler/Y4hSX/src/execution.jl:151
 [11] macro expansion
    @ ~/.julia/packages/CUDA/Tl08O/src/compiler/execution.jl:369 [inlined]
 [12] macro expansion
    @ ./lock.jl:267 [inlined]
 [13] cufunction(f::WaterLily.var"#gpu_##kern_#258#38", tt::Type{Tuple{KernelAbstractions.CompilerMetadata{KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.DynamicCheck, Nothing, CartesianIndices{3, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}, Base.OneTo{Int64}}}, KernelAbstractions.NDIteration.NDRange{3, KernelAbstractions.NDIteration.DynamicSize, KernelAbstractions.NDIteration.StaticSize{(64, 1, 1)}, CartesianIndices{3, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}, Base.OneTo{Int64}}}, Nothing}}, CUDA.CuDeviceArray{Float32, 4, 1}, Int64, Tuple{Int64, Int64, Int64}, CartesianIndex{3}}}; kwargs::@Kwargs{always_inline::Bool, maxthreads::Int64})
    @ CUDA ~/.julia/packages/CUDA/Tl08O/src/compiler/execution.jl:364
 [14] macro expansion
    @ ~/.julia/packages/CUDA/Tl08O/src/compiler/execution.jl:112 [inlined]
 [15] (::KernelAbstractions.Kernel{CUDA.CUDAKernels.CUDABackend, KernelAbstractions.NDIteration.StaticSize{(64,)}, KernelAbstractions.NDIteration.DynamicSize, WaterLily.var"#gpu_##kern_#258#38"})(::CuArray{Float32, 4, CUDA.DeviceMemory}, ::Vararg{Any}; ndrange::Tuple{Int64, Int64, Int64}, workgroupsize::Nothing)
    @ CUDA.CUDAKernels ~/.julia/packages/CUDA/Tl08O/src/CUDAKernels.jl:103
 [16] (::WaterLily.var"##kern#257#36"{Tuple{Int64, Int64, Int64}, Int64, Int64})(a::CuArray{Float32, 4, CUDA.DeviceMemory}, i::Int64, A::Tuple{Int64, Int64, Int64}, ::Val{8})
    @ WaterLily /mnt/c/Users/Brian/Documents/GitHub/WaterLily.jl/src/util.jl:114
 [17] BC!(a::CuArray{Float32, 4, CUDA.DeviceMemory}, A::Tuple{Int64, Int64, Int64}, saveexit::Bool, perdir::Tuple{})
    @ WaterLily /mnt/c/Users/Brian/Documents/GitHub/WaterLily.jl/src/util.jl:116
 [18] Flow(N::Tuple{Int64, Int64, Int64}, U::Tuple{Int64, Int64, Int64}; f::Type, Δt::Float64, ν::Float64, g::Nothing, uλ::Function, perdir::Tuple{}, exitBC::Bool, T::Type)
    @ WaterLily /mnt/c/Users/Brian/Documents/GitHub/WaterLily.jl/src/Flow.jl:115
 [19] Simulation(dims::Tuple{Int64, Int64, Int64}, u_BC::Tuple{Int64, Int64, Int64}, L::Float64; Δt::Float64, ν::Float64, g::Nothing, U::Int64, ϵ::Int64, perdir::Tuple{}, uλ::Function, exitBC::Bool, body::WaterLily.NoBody, T::Type, mem::Type)
    @ WaterLily /mnt/c/Users/Brian/Documents/GitHub/WaterLily.jl/src/WaterLily.jl:74
 [20] tgv(p::Int64, backend::Type; Re::Int64, T::Type)
    @ Main /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:17
 [21] add_to_suite!(suite::BenchmarkGroup, sim_function::typeof(tgv); p::Tuple{Int64, Int64}, s::Int64, ft::Type, backend::Type, bstr::String, remeasure::Bool)
    @ Main /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/util.jl:31
 [22] run_benchmarks(cases::Vector{String}, log2p::Vector{Tuple{Int64, Int64}}, max_steps::Vector{Int64}, ftype::Vector{DataType}, backend::Type, bstr::String; data_dir::SubString{String})
    @ Main /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:72
 [23] top-level scope
    @ /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:87
in expression starting at /mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/benchmark.jl:87
error in running finalizer: CUDA.CuError(code=CUDA.cudaError_enum(0x000002bc))
throw_api_error at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:30
check at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:37 [inlined]
cuModuleUnload at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/utils/call.jl:34 [inlined]
#954 at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/module.jl:92 [inlined]
#context!#990 at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/state.jl:168 [inlined]
context! at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/state.jl:163 [inlined]
unsafe_unload! at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/module.jl:91
jfptr_unsafe_unloadNOT._13453 at /home/mingels28/.julia/compiled/v1.10/CUDA/oWw5k_L5uCo.so (unknown line)
_jl_invoke at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:2895 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:3077
run_finalizer at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gc.c:318
jl_gc_run_finalizers_in_list at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gc.c:408
run_finalizers at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gc.c:454
ijl_atexit_hook at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/init.c:299
ijl_exit at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/init.c:207
exit at ./initdefs.jl:28 [inlined]
exec_options at ./client.jl:323
_start at ./client.jl:552
jfptr__start_82729.1 at /home/mingels28/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:2895 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:3077
jl_apply at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/julia.h:1982 [inlined]
true_main at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/jlapi.c:582
jl_repl_entrypoint at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/jlapi.c:731
main at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/cli/loader_exe.c:58
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
WARNING: Error while freeing DeviceMemory(3.290 MiB at 0x0000000f00000000):
CUDA.CuError(code=CUDA.cudaError_enum(0x000002bc))

Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:30
  [2] check
    @ ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:37 [inlined]
  [3] cuMemFreeAsync
    @ ~/.julia/packages/CUDA/Tl08O/lib/utils/call.jl:34 [inlined]
  [4] free(mem::CUDA.DeviceMemory; stream::CUDA.CuStream)
    @ CUDA ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/memory.jl:87
  [5] free
    @ ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/memory.jl:82 [inlined]
  [6] #1102
    @ ~/.julia/packages/CUDA/Tl08O/src/memory.jl:708 [inlined]
  [7] #context!#990
    @ ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/state.jl:168 [inlined]
  [8] context!
    @ ~/.julia/packages/CUDA/Tl08O/lib/cudadrv/state.jl:163 [inlined]
  [9] _pool_free
    @ ~/.julia/packages/CUDA/Tl08O/src/memory.jl:707 [inlined]
 [10] macro expansion
    @ ./timing.jl:395 [inlined]
 [11] pool_free(managed::CUDA.Managed{CUDA.DeviceMemory})
    @ CUDA ~/.julia/packages/CUDA/Tl08O/src/memory.jl:689
 [12] release(::GPUArrays.RefCounted{CUDA.Managed{CUDA.DeviceMemory}})
    @ GPUArrays ~/.julia/packages/GPUArrays/qt4ax/src/host/abstractarray.jl:42
 [13] unsafe_free!
    @ ~/.julia/packages/GPUArrays/qt4ax/src/host/abstractarray.jl:91 [inlined]
 [14] unsafe_free!(xs::CuArray{Float32, 4, CUDA.DeviceMemory})
    @ CUDA ~/.julia/packages/CUDA/Tl08O/src/array.jl:94
 [15] exit
    @ ./initdefs.jl:28 [inlined]
 [16] exec_options(opts::Base.JLOptions)
    @ Base ./client.jl:323
 [17] _start()
    @ Base ./client.jl:552
error in running finalizer: CUDA.CuError(code=CUDA.cudaError_enum(0x000002bc))
throw_api_error at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:30
check at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/libcuda.jl:37 [inlined]
cuStreamDestroy_v2 at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/utils/call.jl:34 [inlined]
#922 at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/stream.jl:89
#context!#990 at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/state.jl:168
context! at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/state.jl:163
unsafe_destroy! at /home/mingels28/.julia/packages/CUDA/Tl08O/lib/cudadrv/stream.jl:88
jfptr_unsafe_destroyNOT._13181 at /home/mingels28/.julia/compiled/v1.10/CUDA/oWw5k_L5uCo.so (unknown line)
_jl_invoke at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:2895 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:3077
run_finalizer at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gc.c:318
jl_gc_run_finalizers_in_list at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gc.c:408
run_finalizers at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gc.c:454
ijl_atexit_hook at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/init.c:299
ijl_exit at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/init.c:207
exit at ./initdefs.jl:28 [inlined]
exec_options at ./client.jl:323
_start at ./client.jl:552
jfptr__start_82729.1 at /home/mingels28/.julia/juliaup/julia-1.10.4+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
_jl_invoke at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:2895 [inlined]
ijl_apply_generic at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/gf.c:3077
jl_apply at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/julia.h:1982 [inlined]
true_main at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/jlapi.c:582
jl_repl_entrypoint at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/src/jlapi.c:731
main at /cache/build/builder-amdci4-0/julialang/julia-release-1-dot-10/cli/loader_exe.c:58
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
```

**Comment by b-fg** [comment]
The error points at the `BC!` kernel launched when creating `Flow`. Can you double check you can run a simulation on the GPU after installing `nvidia-cuda-toolkit`?

**Comment by mingels28** [comment]
I made a simple example of a moving 2D circle which ran no problem. I called it with ```julia --threads=auto --project TwoD_Circle.jl``` in the same directory as the benchmarking scripts and added a flag for ```@assert CUDA.functional()``` which passed.

**Comment by b-fg** [comment]
Can you please show the example? It seems that `@assert CUDA.functional()` is working with or without `nvidia-cuda-toolkit`.

**Comment by mingels28** [comment]
```julia
using WaterLily, CUDA, StaticArrays
@assert CUDA.functional();

# CUDA.allowscalar(true);

""" geometry function """
function circle(p=8;Re=4000,U=1,Vr=7.2,Ay_D=0.4,Ax_D=0.0,ϕ=0,mem=Array,ϵ=0.1)
    n = 2^p                     #number of grid points
    L = n/20                    #characteristic geometry size
    center = SA[n/2,n/2]        #center of 2D domain    

    #define norm2 function for convenience
    norm2(x) = √sum(abs2,x)
    
    #circle SDF
    function sdf(x,t)
        norm2(x.-center)-L/2
    end

    function map(x,t)
        ω = 2π*U/(Vr*L)
        x .- SA[Ax_D*L*sin(2*ω*t + ϕ), Ay_D*L*sin(ω*t)]
    end

    #2D simulation
    WaterLily.Simulation((2n,n),(U,0),L;U,ν=U*L/Re,
        body=WaterLily.AutoBody(sdf,map),mem=mem,T=Float32,ϵ=ϵ) 
end


""" define the parameters for your simulation """
Vr    = 7.2         #reduced velocity
Ay_D  = 0.35        #CF scaled amplitude
Ax_D  = 0.2         #IL scaled amplitude
phase = 90*π/180   #CF to IL phase offset

p        = 8        #effective grid resolution
n        = 2^p;
Re       = 7620     #Reynolds number
step     = 0.05    #time step for recording data
duration = 20     #length of simulation


""" initialize the simulation """
sim = circle(p;Re=Re,Vr=Vr,Ay_D=Ay_D,Ax_D=Ax_D,ϕ=phase,mem=CUDA.CuArray);
sim_step!(sim,duration,remeasure=true,verbose=true);
```

**Comment by b-fg** [comment]
Ok so indeed is using the GPU correctly in this test. Hence might be a benchmark-test related error for you particular configuration. What CUDA version are you using? If you could go to the WaterLily-Benchmarks project REPL and report the output of `]st` it would be great. Also, you can try running the first TGV benchmark in a separate script, like the one you just shared and see if that works.

**Comment by mingels28** [comment]
Here is the output of ```]st``` my CUDA version definitely looks a bit dated, so I can try updating it to a newer version.

```
(WaterLily-Benchmarks) pkg> st
Status `/mnt/c/Users/Brian/Documents/GitHub/WaterLily-Benchmarks/Project.toml`
  [21141c5a] AMDGPU v1.0.1
  [6e4b80f9] BenchmarkTools v1.5.0
  [052768ef] CUDA v5.4.3
  [324d7699] CategoricalArrays v0.10.8
  [35d6a980] ColorSchemes v3.26.0
  [0c68f7d7] GPUArrays v10.3.1
  [c27321d9] Glob v1.3.1
  [63c18a36] KernelAbstractions v0.9.25
  [b964fa9f] LaTeXStrings v1.3.1
  [91a5bcdd] Plots v1.40.8
  [08abe8d2] PrettyTables v2.3.2
  [90137ffa] StaticArrays v1.9.7
  [f3b207a7] StatsPlots v0.15.7
  [ed894a53] WaterLily v1.2.0 `/mnt/c/Users/Brian/Documents/GitHub/WaterLily.jl`
  ```
  
  and the tgv case, it ran no problem (I commented out the plotting routine and sim_step! for 10 cycles on the GPU)

**Comment by b-fg** [comment]
So that's very weird. Are you running the same size for the TGV?

**Comment by mingels28** [comment]
I only tried ```pow=6``` and ```pow=8``` but both worked no problem. The circle case also ran at 8.

**Comment by mingels28** [comment]
I have to backup a couple steps actually: I ran tgv on it's own and not using ```run_benchmark``` in ```benchmark.jl```. That will still not run due to an illegal memory address, same as before. Calling tgv separately such as with one of the older WaterLily examples works fine though.

**Comment by b-fg** [comment]
Alright, still it's very hard for me to help further without being able to reproduce this. I would suggest to start from scratch in a clean environment:

```sh
git clone https://github.com/WaterLily-jl/WaterLily-Benchmarks
cd WaterLily-Benchmarks
julia --project
]update
# exit Julia REPL
sh benchmark.sh
```

**Comment by mingels28** [comment]
I think I figured it out after re-cloning the benchmark suite and latest WaterLily version in a new environment. I was able to run the benchmark with the GPU on my personal computer using ```bash benchmark.sh -wd "../WaterLily.jl/" -b "CuArray" -c "jelly tgv" -p "5 6"```. I think the main issue was not having the nvidia cuda-toolkit installed which caused ```CuArray``` to not be recognized, and the illegal memory address was caused by running out of VRAM during troubleshooting by having other programs open and running during benchmarking. I don't have an AMD GPU to test on but I assume there would be a similar think with some sort of hipcc toolkit as well. This was all done in WSL as well as in Ubuntu on an HPC cluster.
